src/main/kotlin/com/openlattice/users/Auth0SyncService.kt

```diff
@@ -64,6 +53,34 @@ class Auth0SyncService(
         syncUserEnrollmentsAndAuthentication(user)
     }
 
+    fun createOrUpdateUsers(usersToUpdate: Collection<User>) {
+
+        val usersByPrincipal = usersToUpdate.associateBy { getPrincipal(it) }.toMutableMap()
+
+        val existingUserPrincipals = spm.getSecurablePrincipals(usersByPrincipal.keys).map { it.principal }
+        val principalsToCreate = usersByPrincipal.keys - existingUserPrincipals
+
+        logger.debug("Creating users {} and updating principals {}",
+                principalsToCreate.map { it.id },
+                existingUserPrincipals.map { it.id }
+        )
+
+        val newlyCreatedUsers = principalsToCreate.filter {
+            val wasSuccessfullyCreated = tryCreateNewUserPrincipal(usersByPrincipal.getValue(it), it)
+
+            if (!wasSuccessfullyCreated) {
```

Comment:
This doesn't seem fail safe. If datastore is killed before the body of the if executes, will it be left in an inconsistent state?

Reply:
It would, and I have to double check but I'm pretty sure the auth0 initialization task still runs through every user, so every fresh restart should be a hard re-sync

Reply:
:+1: